SearchRepos is hitting rate limits

**Describe the bug**

```
2024-05-13T17:04:27.0321574Z Starting 1 queries
2024-05-13T17:04:27.8388184Z Search: vulnerability poc pushed:2024-05-12..2024-05-13
2024-05-13T17:04:28.1586607Z Traceback (most recent call last):
2024-05-13T17:04:28.1596262Z   File "/opt/hostedtoolcache/Python/3.9.19/x64/bin/search_github", line 211, in <module>
2024-05-13T17:04:28.1597379Z     main(search_str, args.start_date, args.end_date, args.overwrite)
2024-05-13T17:04:28.1598433Z   File "/opt/hostedtoolcache/Python/3.9.19/x64/bin/search_github", line 56, in main
2024-05-13T17:04:28.1599312Z     data = do_search(query, start_date, end_date)
2024-05-13T17:04:28.1600814Z   File "/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/site-packages/labyrinth/search.py", line 75, in do_search
2024-05-13T17:04:28.1601900Z     for r in result:
2024-05-13T17:04:28.1603074Z   File "/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/site-packages/github/PaginatedList.py", line 84, in __iter__
2024-05-13T17:04:28.1604200Z     newElements = self._grow()
2024-05-13T17:04:28.1605455Z   File "/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/site-packages/github/PaginatedList.py", line 95, in _grow
2024-05-13T17:04:28.1606921Z     newElements = self._fetchNextPage()
2024-05-13T17:04:28.1608379Z   File "/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/site-packages/github/PaginatedList.py", line 244, in _fetchNextPage
2024-05-13T17:04:28.1609730Z     headers, data = self.__requester.requestJsonAndCheck(
2024-05-13T17:04:28.1611192Z   File "/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/site-packages/github/Requester.py", line 548, in requestJsonAndCheck
2024-05-13T17:04:28.1612793Z     return self.__check(*self.requestJson(verb, url, parameters, headers, input, self.__customConnection(url)))
2024-05-13T17:04:28.1614422Z   File "/opt/hostedtoolcache/Python/3.9.19/x64/lib/python3.9/site-packages/github/Requester.py", line 609, in __check
2024-05-13T17:04:28.1615639Z     raise self.createException(status, responseHeaders, data)
2024-05-13T17:04:28.1619712Z github.GithubException.RateLimitExceededException: 403 {"documentation_url": "https://docs.github.com/free-pro-team@latest/rest/overview/rate-limits-for-the-rest-api#about-secondary-rate-limits", "message": "You have exceeded a secondary rate limit. Please wait a few minutes before you try again. If you reach out to GitHub Support for help, please include the request ID B470:3F0D3F:14DBB1E9:21B8E305:6642481B."}
2024-05-13T17:04:28.2168365Z ##[error]Process completed with exit code 1.
```

**To Reproduce**

Steps to reproduce the behavior:
1. Go to https://github.com/CERTCC/labyrinth/actions/workflows/repo_search.yml
2. Observe that SearchRepos sometimes fails due to the above error

**Expected behavior**

It should be able to catch the secondary rate limit failure, sleep for a few minutes and try again with an exponential backoff or something.

--- a/labyrinth/search.py
+++ b/labyrinth/search.py
@@ -8,6 +8,9 @@ import datetime
 
 import pandas as pd
 from github import Github
+from github.GithubException import RateLimitExceededException
+import random
+import time
 
 import labyrinth
 from labyrinth.date_helpers import fixup_start_date, fixup_end_date
@@ -63,10 +66,22 @@ def do_search(query, start_date=None, end_date=None, page_size=100):
 
         # result is a generator of repository objects
         count = 0
-        for r in result:
-            # check your rate limits every so often
-            if len(results) % 50 == 0:
+
+        # manually iterate over the generator so we can catch rate limit exceptions at
+        # an individual result level
+        while True:
+            try:
+                r = next(result)
+            except StopIteration:
+                break
+            except RateLimitExceededException as e:
+                # we need to catch github.GithubException.RateLimitExceededException
+                print("Rate limit exceeded")
+                # take a random nap
+                time.sleep(random.randint(10, 30))
+                # and check rate limits
                 check_rate_limits(gh)
+                continue
 
             # timestamp
             ts = (
@@ -82,6 +97,7 @@ def do_search(query, start_date=None, end_date=None, page_size=100):
             data["matched_at"] = ts
             results.append(data)
             count += 1
+
         print(f"Found {count} results for {qstr}", flush=True)
 
     return results

